cpp/src/barretenberg/dsl/acir_format/recursion_constraint.cpp

```diff
@@ -0,0 +1,347 @@
+            // This will make the RecursionConstraint idx equal to `dummy_field`.
+            // In the case of a valid witness assignment, this does nothing (as dummy_field = real value)
+            // In the case of no valid witness assignment, this makes sure that the RecursionConstraint witness indices
+            // will not trigger basic errors (check inputs are on-curve, check we are not inverting 0)
+            composer.assert_equal(composer.add_variable(dummy_field), proof_field_idx);
+        }
+        for (size_t i = 0; i < input.key.size(); ++i) {
+            const auto key_field_idx = input.key[i];
+            barretenberg::fr dummy_field =
+                has_valid_witness_assignments ? composer.get_variable(key_field_idx) : dummy_key[i];
+            composer.assert_equal(composer.add_variable(dummy_field), key_field_idx);
+        }
+    }
+
+    // Construct an in-circuit representation of the verification key.
+    // For now, the v-key is a circuit constant and is fixed for the circuit.
+    // (We may need a separate recursion opcode for this to vary, or add more config witnesses to this opcode)
+    const auto& aggregation_input = input.input_aggregation_object;
+    aggregation_state_ct previous_aggregation;
+
+    // If we have previously recursively verified proofs, `is_aggregation_object_nonzero = true`
+    // For now this is a complile-time constant i.e. whether this is true/false is fixed for the circuit!
+    bool inner_aggregation_indices_all_zero = true;
+    for (const auto& idx : aggregation_input) {
+        inner_aggregation_indices_all_zero &= (idx == 0);
+    }
+    if (!inner_aggregation_indices_all_zero) {
+        std::array<bn254::fq_ct, 4> aggregation_elements;
+        for (size_t i = 0; i < 4; ++i) {
+            aggregation_elements[i] =
+                bn254::fq_ct(field_ct::from_witness_index(&composer, aggregation_input[4 * i]),
+                             field_ct::from_witness_index(&composer, aggregation_input[4 * i + 1]),
+                             field_ct::from_witness_index(&composer, aggregation_input[4 * i + 2]),
+                             field_ct::from_witness_index(&composer, aggregation_input[4 * i + 3]));
+            aggregation_elements[i].assert_is_in_field();
+        }
+        // If we have a previous aggregation object, assign it to `previous_aggregation` so that it is included
+        // in stdlib::recursion::verify_proof
+        previous_aggregation.P0 = bn254::g1_ct(aggregation_elements[0], aggregation_elements[1]);
+        previous_aggregation.P1 = bn254::g1_ct(aggregation_elements[2], aggregation_elements[3]);
+        previous_aggregation.has_data = true;
+    } else {
+        previous_aggregation.has_data = false;
+    }
+
+    transcript::Manifest manifest = Composer::create_unrolled_manifest(input.public_inputs.size());
+
+    std::vector<field_ct> key_fields;
```

Comment:
magic number `1` here. Might be better to replace in this test w. hardcoded `const size_t num_public_inputs = 1`